yarn/src/main/scala/org/apache/spark/deploy/yarn/Client.scala

```diff
@@ -616,17 +620,29 @@ private[spark] class Client(
    * throw an appropriate SparkException.
    */
   def run(): Unit = {
-    val (yarnApplicationState, finalApplicationStatus) = monitorApplication(submitApplication())
-    if (yarnApplicationState == YarnApplicationState.FAILED ||
-      finalApplicationStatus == FinalApplicationStatus.FAILED) {
-      throw new SparkException("Application finished with failed status")
-    }
-    if (yarnApplicationState == YarnApplicationState.KILLED ||
-      finalApplicationStatus == FinalApplicationStatus.KILLED) {
-      throw new SparkException("Application is killed")
-    }
-    if (finalApplicationStatus == FinalApplicationStatus.UNDEFINED) {
-      throw new SparkException("The final status of application is undefined")
+    val appId = submitApplication()
+    if (fireAndForget) {
+      logInfo("... polling ResourceManager for application state")
+      val report = getApplicationReport(appId)
+      val state = report.getYarnApplicationState
+      if (state == YarnApplicationState.FAILED || state == YarnApplicationState.KILLED) {
+        throw new SparkException(s"Application $appId finished with status: $state")
+      }
+      logInfo(s"Application report for $appId (state: $state)")
+      logInfo(formatReportDetails(report))
```

Comment:
I tried this out in a few cases and I noticed if we failed to submit there is no message telling the user why. we need print the report in the case the state failed/killed too.  Make sure the diagnostics message is there.